### PR TITLE
avoid ResourceWarning in `DataBody.__aiter__` 

### LINKED ISSUE #301: ResourceWarning from `DataBody.__aiter__`
from quart-trio I get a:

```ResourceWarning: Async generator 'quart.wrappers.response.DataBody.__aiter__.<locals>._aiter' was garbage collected before it had been exhausted. Surround its use in 'async with aclosing(...):' to ensure that it gets cleaned up as soon as you're done using it.```

Environment:

- Python version:
- Quart version:

### PR DESCRIPTION
**Pull request by graingert**
Avoid a ResourceWarning by avoiding an async generator

- fixes #301

Checklist:

- [ ] Add tests that demonstrate the correct behavior of the change. Tests should fail without the change.
- [ ] Add or update relevant docs, in the docs folder and in code.
- [ ] Add an entry in `CHANGES.rst` summarizing the change and linking to the issue.
- [ ] Add `.. versionchanged::` entries in any relevant code docs.
- [x] Run `pre-commit` hooks and fix any issues.
- [x] Run `pytest` and `tox`, no tests failed.

### CODE REVIEW

**Comment by graingert** [comment]
tests are tricky for this, as it only causes the warning on quart-trio

File: src/quart/wrappers/response.py [hunk 1]
```diff
@@ -91,6 +88,20 @@ async def make_conditional(self, begin: int, end: int | None) -> int:
...
+    def __init__(self, data_body: DataBody):
+        self._data_body = data_body
+        self._iterated = False
+
+    async def __anext__(self) -> bytes:
+        if self._iterated:
+            raise StopAsyncIteration
+
+        self._iterated = True
+        body = self._data_body
```
Comment (davidism) [inline src/quart/wrappers/response.py:101]: This type of optimization should no longer be needed in modern Python.

File: src/quart/wrappers/response.py [hunk 2]
```diff
@@ -91,6 +88,19 @@ async def make_conditional(self, begin: int, end: int | None) -> int:
...
+    def __init__(self, data_body: DataBody):
+        self._data_body = data_body
+        self._iterated = False
+
+    async def __anext__(self) -> bytes:
+        if self._iterated:
+            raise StopAsyncIteration
+
+        self._iterated = True
+        return self._data_body.body[self._data_body.begin : self._data_body.end]
```
Comment (graingert) [inline src/quart/wrappers/response.py:101]:
```suggestion
        return self._data_body.data[self._data_body.begin : self._data_body.end]
```

**Comment by pgjones** [comment]
Thanks

**Pull request merged**